05.ls/ls.rb

```diff
@@ -31,8 +33,133 @@ def list_files(file_names)
   end
 end
 
-if options['r']
-  list_files(Dir.glob('*').reverse)
+def count_blocks(file_names)
+  file_names.map do |fname|
+    fs = File::Stat.new(fname)
+    fs.blocks / 2
+  end.sum
+end
+
+def list_file_type(file_mode)
+  file_type = {
+    '04' => 'd',
+    '10' => '-',
+    '12' => 'l'
+  }
+  type = file_mode[0] + file_mode[1]
+  file_type[type]
+end
+
+def list_special_perm_suid(file_mode, file_permissions)
+  if file_mode[2] == 4 && file_permissions[2] == 'x'
+    file_permissions[2] = 'S'
+  elsif file_mode[2] == 4
+    file_permissions[2] = 's'
+  end
+end
+
+def list_special_perm_sgid(file_mode, file_permissions)
+  if file_mode[2] == 2 && file_permissions[5] == 'x'
+    file_permissions[5] = 'S'
+  elsif file_mode[2] == 2
+    file_permissions[5] = 's'
+  end
+end
+
+def list_special_perm_sticky(file_mode, file_permissions)
+  if file_mode[2] == 1 && file_permissions[8] == 'x'
+    file_permissions[8] = 'T'
+  elsif file_mode[2] == 1
+    file_permissions[8] = 't'
+  end
+end
+
+def list_file_perm(file_mode)
+  file_permissions =
+    (3..5).map do |fp|
+      fm = file_mode[fp].to_i
+      fm_r = fm >= 4 ? 'r' : '-'
+      fm -= 4 if fm >= 4
+      fm_w = fm >= 2 ? 'w' : '-'
+      fm -= 2 if fm >= 2
+      fm_x = fm >= 1 ? 'x' : '-'
+      [fm_r, fm_w, fm_x]
+    end
+  list_special_perm_suid(file_mode, file_permissions)
+  list_special_perm_sgid(file_mode, file_permissions)
+  list_special_perm_sticky(file_mode, file_permissions)
+  file_permissions.join
+end
+
+def get_files_in_long_format(file_names)
+  file_names.map do |fname|
+    fs = File::Stat.new(fname)
+    {
+      file_mode: fs.mode,
+      hard_link: fs.nlink,
+      user_id: fs.uid,
+      group_id: fs.gid,
+      file_size: fs.size,
+      last_update_time: fs.mtime,
+      file_name: fname
+    }
+  end
+end
+
+def convert_files_in_long_format(files_in_long_format)
+  files_in_long_format.map do |longformat|
+    file_mode_octal = longformat[:file_mode].to_s(8).split(//)
+    file_mode_octal.unshift('0') if file_mode_octal.size == 5
+    time_stamp = longformat[:last_update_time].to_a
+    date = Date.new(time_stamp[5], time_stamp[4], time_stamp[3])
+    minute = time_stamp[1].to_s.size == 1 ? ['0', time_stamp[1]].join : time_stamp[1]
+    {
+      type_and_permissions: [list_file_type(file_mode_octal), list_file_perm(file_mode_octal)].join,
+      number_of_hard_links: longformat[:hard_link].to_s,
+      user_name: Etc.getpwuid(longformat[:user_id]).name,
+      group_name: Etc.getgrgid(longformat[:group_id]).name,
+      file_size: longformat[:file_size].to_s,
+      month: date.strftime('%b'),
+      day: time_stamp[3].to_s,
+      time: [time_stamp[2], minute].join(':'),
+      file_name: longformat[:file_name]
+    }
+  end
+end
+
+def line_up_long_format(files_in_long_format)
+  key_of_file_elements = files_in_long_format[0].map { |k, _v| k }
+  max_number_of_chars =
+    key_of_file_elements.map do |kfe|
```

Comment:
改行しない方がインデントが少なくて済むし読みやすいと思います〜
```suggestion
  max_number_of_chars = key_of_file_elements.map do |kfe|
```